### PR TITLE
feat: sampling based allocation

### PR DESCRIPTION
**Pull request by mooselumph**
## Why are these changes needed?

Validators use pseudorandom chunk assignment for bandwidth conservation across quorums. 

## Checks

- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, in that case, please comment that they are not relevant.
- [ ] I've checked the new test coverage and the coverage percentage didn't drop.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Integration tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: api/clients/v2/validator/retrieval_worker.go [hunk 1]
```diff
@@ -186,6 +184,9 @@ type retrievalWorker struct {
 
 	// Counts the number of chunks in each status.
 	chunkStatusCounts map[chunkStatus]int
+
+	// The status of each chunk
+	chunkOwner map[uint32]core.OperatorID
```
Comment (cody-littley) [inline api/clients/v2/validator/retrieval_worker.go:189]:
```suggestion
	// For the purposes of retrieving this blob, the validator that we have decided to mark as being the "owner".
	// Although it's possible that a particular chunk might be available from more than one validator, we only
	// ever attempt to fetch any particular chunk from a single source.
	chunkOwner map[uint32]core.OperatorID
```

**Review by cody-littley — COMMENTED** [review]
I started my review but didn't fully finish. Will have the first round of reviews completed before you start work on Monday.

File: core/v2/assignment_test.go [hunk 2]
```diff
@@ -11,136 +13,55 @@ import (
...
-			RHS := 100 * myStake * blobParams.NumChunks
-
-			assert.GreaterOrEqual(t, LHS, RHS)
-		}
-	})
-}
+var blobKey1 = []byte("blobKey1")
+var blobKey2 = []byte("blobKey2")
+
+// func TestOperatorAssignmentsV2(t *testing.T) {
```
Comment (litt3) [inline core/v2/assignment_test.go:19]: is code intentionally left commented out?

File: core/v2/types.go [hunk 3]
```diff
@@ -83,6 +82,17 @@ type BlobHeader struct {
 	PaymentMetadata core.PaymentMetadata
 }
 
+type BlobHeaderWithoutPayment struct {
```
Comment (litt3) [inline core/v2/types.go:85]: I think `BlobHeaderWithHashedPayment` might be a clearer name for this struct.

Also consider a brief doc, describing why this can/must exist

File: core/data.go [hunk 4]
```diff
@@ -649,9 +649,11 @@ type OnDemandPayment struct {
 }
 
 type BlobVersionParameters struct {
-	CodingRate      uint32
-	MaxNumOperators uint32
-	NumChunks       uint32
+	CodingRate                  uint32
+	ReconstructionThresholdBips uint32
```
Comment (litt3) [inline core/data.go:653]: Perhaps the new fields here could be given some docs?

File: inabox/tests/integration_test.go [hunk 5]
```diff
@@ -27,6 +27,7 @@ func mineAnvilBlocks(numBlocks int) {
 
 var _ = Describe("Inabox Integration", func() {
 	It("test end to end scenario", func() {
+		Skip("Don't run v1 test right now")
```
Comment (litt3) [inline inabox/tests/integration_test.go:30]: Is this meant to merge into `master` skipping v1 tests?

File: node/grpc/server_v2.go [hunk 6]
```diff
@@ -180,35 +180,29 @@ func (s *ServerV2) validateAndStoreChunks(
...
-			bundleKey, err := node.BundleKey(blobKey, quorum)
-			if err != nil {
-				return api.NewErrorInternal("failed to get bundle key")
-			}
-
-			batchData = append(batchData, &node.BundleToStore{
-				BundleKey:   bundleKey,
-				BundleBytes: bundle,
-			})
+		quorum := core.QuorumID(0)
```
Comment (litt3) [inline node/grpc/server_v2.go:196]: Here's another case where I'm not sure what to make of the hardcoded quorum number

File: node/grpc/server_v2.go [hunk 7]
```diff
@@ -342,7 +336,10 @@ func (s *ServerV2) GetChunks(ctx context.Context, in *pb.GetChunksRequest) (*pb.
 	if corev2.MaxQuorumID < in.GetQuorumId() {
 		return nil, api.NewErrorInvalidArg("invalid quorum ID")
 	}
-	quorumID := core.QuorumID(in.GetQuorumId())
+
+	// Ignore the quorum ID and always return the bundle for quorum 0
+	// quorumID := core.QuorumID(in.GetQuorumId())
```
Comment (litt3) [inline node/grpc/server_v2.go:341]: is this line intentionally left commented out?

File: node/node_v2.go [hunk 8]
```diff
@@ -68,52 +68,45 @@ func (n *Node) DownloadBundles(
...
-				return nil, nil, fmt.Errorf("blob version %d not found", cert.BlobHeader.BlobVersion)
-			}
 
-			if _, ok := operatorState.Operators[quorum]; !ok {
-				// operator is not part of the quorum or the quorum is not valid
-				n.Logger.Debug("operator is not part of the quorum or the quorum is not valid",
-					"quorum", quorum)
-				continue
-			}
+		// Old quorum loop started here
```
Comment (litt3) [inline node/node_v2.go:78]: are these "old quorum loop" comments intentionally left in? Seems like the kind of thing that an LLM left behind?

File: node/node_v2.go [hunk 9]
```diff
@@ -163,13 +156,14 @@ func (n *Node) DownloadBundles(
 				len(resp.bundles), len(resp.metadata))
 		}
 
-		for j, bundle := range resp.bundles {
-			metadata := resp.metadata[j]
-			blobShards[metadata.blobShardIndex].Bundles[metadata.quorum], err = new(core.Bundle).Deserialize(bundle)
+		for i, bundle := range resp.bundles {
```
Comment (litt3) [inline node/node_v2.go:159]: incorrect merge happened here, this is undoing one of Cody's audit finding fixes (nested loop index var should be `j`, not `i`)

File: test/v2/client/test_client.go [hunk 10]
```diff
@@ -502,19 +501,20 @@ func (c *TestClient) DisperseAndVerify(ctx context.Context, payload []byte) erro
 		return fmt.Errorf("failed to read blob from relays: %w", err)
 	}
 
-	blobHeader := eigenDACert.BlobInclusionInfo.BlobCertificate.BlobHeader
-	commitment, err := coretypes.BlobCommitmentsBindingToInternal(&blobHeader.Commitment)
+	// blobHeader := eigenDACert.BlobInclusionInfo.BlobCertificate.BlobHeader
```
Comment (litt3) [inline test/v2/client/test_client.go:504]: intentionally left commented out?

File: test/v2/client/test_client.go [hunk 11]
```diff
@@ -661,23 +626,21 @@ func (c *TestClient) ReadBlobFromValidatorsInQuorum(
 		defer cancel()
 	}
 
+	start := time.Now()
```
Comment (litt3) [inline test/v2/client/test_client.go:629]: this change looks like another merge mistake, presumably `start` should be defined inside the `validateAndDecode` block

File: test/v2/client/test_client.go [hunk 12]
```diff
@@ -661,23 +626,21 @@ func (c *TestClient) ReadBlobFromValidatorsInQuorum(
...
-			uint64(referenceBlockNumber),
-			quorumID)
+			header,
+			uint64(referenceBlockNumber))
 		if err != nil {
 			return fmt.Errorf("failed to read blob from validators, %s", err)
 		}
 
-		c.metrics.reportValidatorReadTime(time.Since(start), quorumID)
+		c.metrics.reportValidatorReadTime(time.Since(start), 0)
```
Comment (litt3) [inline test/v2/client/test_client.go:641]: optionally, you could remove the `quorumID` parameter from `reportValidatorReadTime` entirely, since this is the only usage of that method, and `quorumID` is no longer relevant

File: test/v2/load/load_generator.go [hunk 13]
```diff
@@ -293,27 +292,18 @@ func (l *LoadGenerator) readBlobFromValidators(
 		l.metrics.endOperation("validator_read")
 	}()
 
-	blobHeader := eigenDACert.BlobInclusionInfo.BlobCertificate.BlobHeader
-	commitment, err := coretypes.BlobCommitmentsBindingToInternal(&blobHeader.Commitment)
+	blobHeader, err := coretypes.BlobHeaderBindingToInternal(&eigenDACert.BlobInclusionInfo.BlobCertificate.BlobHeader)
 	if err != nil {
-		l.metrics.reportValidatorReadFailure()
-		l.client.GetLogger().Errorf("failed to bind blob commitments: %v", err)
+		l.client.GetLogger().Errorf("failed to bind blob header: %v", err)
```
Comment (litt3) [inline test/v2/load/load_generator.go:297]: did you intentionally remove the `l.metrics.reportValidatorReadFailure()` call inside this error case?

File: node/node_v2_test.go [hunk 14]
```diff
@@ -60,36 +52,11 @@ func TestDownloadBundles(t *testing.T) {
...
-	require.Contains(t, blobShards[1].Bundles, core.QuorumID(0))
-	require.Contains(t, blobShards[1].Bundles, core.QuorumID(1))
-	require.Contains(t, blobShards[2].Bundles, core.QuorumID(1))
-	require.Contains(t, blobShards[2].Bundles, core.QuorumID(2))
-	bundleEqual(t, bundles[0][0], blobShards[0].Bundles[0])
-	bundleEqual(t, bundles[0][1], blobShards[0].Bundles[1])
-	bundleEqual(t, bundles[1][0], blobShards[1].Bundles[0])
-	bundleEqual(t, bundles[1][1], blobShards[1].Bundles[1])
-	bundleEqual(t, bundles[2][1], blobShards[2].Bundles[1])
-	bundleEqual(t, bundles[2][2], blobShards[2].Bundles[2])
```
Comment (litt3) [inline node/node_v2_test.go:55]: You've removed all usages of `bundleEqual` and `frameEqual`, which could now be deleted

File: api/clients/v2/validator/retrieval_worker.go [hunk 15]
```diff
@@ -257,12 +256,22 @@ func newRetrievalWorker(
...
 	chunkStatusCounts := make(map[chunkStatus]int)
+
 	for _, opID := range downloadOrder {
+		for _, index := range assignments[opID].Indices {
+			_, ok := chunkOwner[index]
+			if !ok {
+				chunkStatusCounts[available]++
+				chunkOwner[index] = opID
+			}
+		}
```
Comment (cody-littley) [inline api/clients/v2/validator/retrieval_worker.go:269]: Although we don't need to assign an owner in the constructor [as discussed in this thread](https://github.com/Layr-Labs/eigenda/pull/1552/files#r2082444944), it is necessary to identify chunks with no possible owners. These chunks should automatically be considered "failed" (by incrementing `chunkStatusCounts[failed]`), since there is no way to acquire these chunks.

This is important. The code keeps track of the number of failed chunks so that it knows when the blob is impossible to reconstruct. In the extreme case where all validators somehow choose the same subset of chunks and most chunks are not owned, then retrieval will always fail. Without marking unowned blobs as failed, I'd expect this utility to block forever (or until the parent context is cancelled).

File: api/clients/v2/validator/validator_client_test.go [hunk 16]
```diff
@@ -93,9 +133,21 @@ func TestBasicWorkflow(t *testing.T) {
 		// only permit downloads to happen once per operator
 		_, ok = downloadSet.Load(operatorID)
 		require.False(t, ok)
+
 		downloadSet.Store(operatorID, struct{}{})
 
-		chunksDownloaded.Add(uint32(len(chunks)))
+		// TODO(Claude): Please apply this de-duplication logic to the other tests as well. It also needs to be performed for the
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply api/clients/v2/validator/validator_client_test.go:139]: Somebody is going to be working on something unrelated, and be super confused why claude suddenly starts making changes to the validator client. 🤣

File: api/clients/v2/validator/retrieval_worker.go (hunk 15 above)
Comment (cody-littley) [reply api/clients/v2/validator/retrieval_worker.go:269]: Probably also worthwhile adding a unit test for this scenario.

File: api/clients/v2/validator/retrieval_worker.go [hunk 17]
```diff
@@ -305,14 +314,25 @@ func newRetrievalWorker(
...
-	operatorChunks := w.assignments[operatorID].NumChunks
 
 	oldStatus, ok := w.chunkStatusMap[operatorID]
+	w.chunkStatusMap[operatorID] = status
+
+	numChunks := 0
+	for _, index := range w.assignments[operatorID].Indices {
+		_, ok := w.chunkOwner[index]
+		if !ok {
+			w.chunkOwner[index] = operatorID
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply api/clients/v2/validator/retrieval_worker.go:325]: Yes, but only if the chunk has no possible owner (more on that in a later comment).

The code assigns an owner both here and in the function that sets status. We can probably remove the status change code from the other place, since I don't think it's possible for that code to trigger.

File: node/node_v2.go (hunk 8 above)
Comment (mooselumph) [reply node/node_v2.go:78]: I wrote these. 😁 Will clean up

File: node/node_v2.go (hunk 9 above)
Comment (mooselumph) [reply node/node_v2.go:159]: Thanks for catching this!

File: test/v2/load/load_generator.go (hunk 13 above)
Comment (mooselumph) [reply test/v2/load/load_generator.go:297]: Not sure what happened here. Thanks for catching.

Comment (mooselumph) [reply test/v2/load/load_generator.go:297]: I replaced this entire block of code with a different function call and forgot to add this metrics call. Thanks for catching.

File: api/clients/v2/validator/retrieval_worker.go [hunk 18]
```diff
@@ -186,6 +184,9 @@ type retrievalWorker struct {
 
 	// Counts the number of chunks in each status.
 	chunkStatusCounts map[chunkStatus]int
+
+	// The status of each chunk
```
[the remark this answers is not in the document]
Comment (mooselumph) [reply api/clients/v2/validator/retrieval_worker.go:188]: I think it's sufficient to only utilize the owner to filter the updates. 
Currently, there's an insignificant overhead from downloading and verifying these additional chunks, and then essentially discarding them. 
This avoids a lot of extra changes which would be entailed by propagating the filtering to the other parts of the code. 

I'll add a unit test to assert that recovery still works when there are duplicate copies of a chunk.

File: api/clients/v2/validator/retrieval_worker.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (mooselumph) [reply api/clients/v2/validator/retrieval_worker.go:325]: The code path here should never be activated, so we could technically remove it.

File: api/clients/v2/validator/retrieval_worker.go [hunk 19]
```diff
@@ -186,6 +183,15 @@ type retrievalWorker struct {
 
 	// Counts the number of chunks in each status.
 	chunkStatusCounts map[chunkStatus]int
+
+	// For the purposes of retrieving this blob, the validator that we have decided to mark as being the "owner".
```
Comment (mooselumph) [inline api/clients/v2/validator/retrieval_worker.go:187]: Updated the documentation around how redundant chunks are handled.

File: api/clients/v2/validator/retrieval_worker.go (hunk 1 above)
Comment (mooselumph) [reply api/clients/v2/validator/retrieval_worker.go]: Updated comment here: https://github.com/Layr-Labs/eigenda/pull/1552/files#r2085270838

File: api/clients/v2/validator/retrieval_worker.go [hunk 20]
```diff
@@ -546,7 +567,7 @@ func (w *retrievalWorker) downloadChunks(operatorID core.OperatorID) {
 	ctx, cancel := context.WithTimeout(w.downloadAndVerifyCtx, w.config.DownloadTimeout)
 	defer cancel()
 
-	reply, err := w.validatorGRPCManager.DownloadChunks(ctx, w.blobKey, operatorID, w.quorumID)
+	reply, err := w.validatorGRPCManager.DownloadChunks(ctx, w.blobKey, operatorID, 0)
```
[the remark this answers is not in the document]
Comment (mooselumph) [reply api/clients/v2/validator/retrieval_worker.go]: I updated the function signature and made some fixes so that the quorumID is irrelevant.

**Review by cody-littley — APPROVED** [review]

File: encoding/rs/decode.go [hunk 21]
```diff
@@ -24,16 +24,27 @@ func (e *Encoder) Decode(frames []FrameCoeffs, indices []uint64, maxInputSize ui
...
 	}
 
-	numSys := encoding.GetNumSys(maxInputSize, g.ChunkLength)
+	if len(frames) != len(indices) {
+		return nil, errors.New("number of frames must equal number of indices")
+	}
+
+	// Remove duplicates
+	frameMap := make(map[uint64]FrameCoeffs, len(indices))
+	for i, index := range indices {
```
Comment (litt3) [inline encoding/rs/decode.go:33]: nit: having loop variables named both `i` and `index` is confusing- consider renaming `index` to something like `frameIndex` to help remove ambiguity

File: retriever/v2/server.go [hunk 22]
```diff
@@ -72,13 +72,15 @@ func (s *Server) RetrieveBlob(ctx context.Context, req *pb.BlobRequest) (*pb.Blo
 	ctxWithTimeout, cancel := context.WithTimeout(ctx, s.config.Timeout)
 	defer cancel()
 
+	BlobHeaderWithHashedPayment, err := blobHeader.GetBlobHeaderWithHashedPayment()
```
Comment (litt3) [inline retriever/v2/server.go:75]: nit: `BlobHeaderWithHashedPayment` should be `blobHeaderWithHashedPayment`

**Review by litt3 — DISMISSED** [review]

File: api/clients/v2/validator/validator_client.go [hunk 23]
```diff
@@ -148,3 +165,16 @@ func (c *validatorClient) GetBlob(
 	}
 	return data, nil
 }
+
+func GetFlattenedOperatorSockets(operatorsMap map[core.QuorumID]map[core.OperatorID]*core.OperatorInfo) map[core.OperatorID]core.OperatorSocket {
```
[the remark this answers is not in the document]
Comment (mooselumph) [reply api/clients/v2/validator/validator_client.go:169]: Added method doc

**Review by cody-littley — APPROVED** [review]
